### PR TITLE
RSDK-6712 APP-3548 meta.json overrides section

### PR DESCRIPTION
**Pull request by abe-winter**
## What changed
- add `overrides` section to meta.json; this edits some meta.json fields based on platform constraints. look at module.schema.json for examples
    - note: meta.json without overrides will work as before
- a 'dev' tag lets users choose overrides when using local + remote reloading
- document overrides in module.schema.json
- important and subtle: for remote reloading, I'm saving the post-override meta.json into the tarball
- new tarball helper in utils
## Why
- the ability to do platform overrides is a frequent user request
- for python reloading users, this lets people automatically toggle between virtualenv (local + remote reloading) and pyinstaller (publishing)
## Design questions
- ~are people happy with the `linux/arm64;dev` format for the override keys? as we evolve our platform tagging this may eventually have a v2, but is this good for now?~ -> going with simpler approach that doesn't combine tags; can add it later if needed
## Testing / sample
Local and remote reloading work as expected with this config:

commands:
```sh
$ ./viam module reload --local --no-build
$ ./viam module reload --home $HOME
```

meta.json:
```json
{
	"$schema": "https://dl.viam.dev/module.schema.json",
	"module_id": "test:test",
	"entrypoint": "dist/main",
	"build": {
		"build": "make dist/main",
		"path": "dist/main"
	},
	"overrides": {
		";dev": {
			"entrypoint": "run.sh",
			"build": {"build": "make module.tar.gz", "path": "module.tar.gz"}
		}
	}
}
```
run.sh:
```bash
#!/usr/bin/env bash

set -e

uv v
uv pip install -r requirements.txt

.venv/bin/python main.py $@
```

### CODE REVIEW

**Comment by abe-winter** [comment]
reviewers: this one is semi painful to read

start reading the module.schema.json changes, should give a good overview

can take you through it live if easier

**Review by zaporter-work — CHANGES_REQUESTED** [review]
Talked offline about override format. Suggested separating out the dev env from the platform. 

Ex:

overrides:
"dev": {some overrides}
"linux/amd64": {other overrides}


This will lead to some duplication but will be more user-understandable

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:657]: outdated comment?

Dev means dev environment now, right?

File: cli/utils.go
Comment (zaporter-work) [inline cli/utils.go:128]: any reason all of the prior logic is wrapped in a function if we are just going to return the err directly?

I suspect you did this for cleanup reasons but I don't see it.

File: cli/module_build.go
Comment (zaporter-work) [inline cli/module_build.go:59]: [aside] Could be cool to have a `--dev` flag to the build cmd to change this behavior.

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:679]: This is a super cool function. I really like this. `build.setup` is cool

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:83]:
```
➜  tmp git:(meta-platform-overrides) ✗ viam module create --public-namespace foo --name baz --local-only
Configuration for the module has been written to meta.json
➜  tmp git:(meta-platform-overrides) ✗ cat meta.json
{
  "$schema": "https://dl.viam.dev/module.schema.json",
  "module_id": "foo:baz",
  "visibility": "private",
  "url": "",
  "description": "",
  "models": [
    {
      "api": "",
      "model": ""
    }
  ],
  "entrypoint": "",
  "overrides": null
}%                                                                                                                                                                                         
```

Suggestion:

Either add omitempty to this or set the default to an empty map in the creation logic. (so that it is not `null`)

**Review by zaporter-work — APPROVED** [review]
Code + features LGTM % @michaellee1019 's review.

I /have not tested this manually/.

File: cli/module_build.go
Comment (zaporter-work) [reply cli/module_build.go:59]: (we should probably leave that to a followup pr)

File: cli/utils.go
Comment (abe-winter) [reply cli/utils.go:128]: I wanted to make sure `defer sourceFile.Close()` ran before the rename. in practice I'm not sure this matters, but it's the kind of thing that can be platform specific so I don't trust my testing to be general

should I comment / am I overthinking this?

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go:59]: yup followup

File: cli/module_registry.go
Comment (abe-winter) [reply cli/module_registry.go:679]: :bow:

File: cli/module_registry.go
Comment (abe-winter) [reply cli/module_registry.go:657]: good question -- I didn't think carefully about how your tagging change affects the application of tags

the current behavior is to apply every override section that matches. so if someone had a `linux/any` and a `dev`, the linux/any would apply even in dev mode. the override sections would get applied in order and where there's a conflict, the latest would rule. (assuming map order is consistent).

if we change, will someone with a linux override say to themselves 'I'm on linux why isn't my linux override running'?

File: cli/utils.go
Comment (zaporter-work) [reply cli/utils.go:128]: Ah -- that makes sense. LGTM.

A comment could be good. But I also doubt anyone will change this.

Comment (abe-winter) [reply cli/utils.go:128]: commented in 898070b7d

File: cli/module_registry.go
Comment (zaporter-work) [reply cli/module_registry.go:657]: > if someone had a linux/any and a dev, the linux/any would apply even in dev mode.

I think this makes sense if the dev overrides are applied after.

Consider if macOS had a different entrypoint, even if I changed the build step in `dev`, I would want the mac overrides to apply (to fix entrypoint) and then I would also want the dev overrides to stack on top to change the build / makefile / etc. 


 This is a product decision and you are better at those than I am

Comment (zaporter-work) [reply cli/module_registry.go:657]: I think if we decide the correct order to apply the changes (sort linux -> linux/arm -> dev) then we would get intuitive behavior.

Comment (abe-winter) [reply cli/module_registry.go:657]: hmm map order is explicitly not preserved in json parsing https://github.com/golang/go/issues/27179

and implicitly seems to be alphabetical, which would give darwin -> dev -> linux :grimacing: 

I think this is going to be a bit surprising no matter what, but your suggestion makes sense -- I'm going to do that + doc in the jsonschema

File: cli/module.schema.json
Comment (michaellee1019) [inline cli/module.schema.json:54]: Sorry that I'm late to reviewing this idea. A few concerns:
- Lets get approval for it (I can send the email) since it does vary from the scope that was approved for adding platform-level commands. 
- I think I prefer `local` over `dev` since that matches the rest of the terminology of running a local module (vs registry module)
- Being able to overwrite _all_ fields feels weird. I think some things shouldn't be overwritten (module_id, description, visibility, etc) and could cause some weird behavior (now and in the future if we depend on meta.json in more places).

File: cli/module.schema.json
Comment (michaellee1019) [inline cli/module.schema.json:48]: This is probably just a random example. If we do indeed want platform specific entrypoints, which makes sense if build scripts vary so much, then:

I think we'd need cloud changes to UpdateModule and the Config APIs. The executable path is constructed on the cloud side from the current entrypoint field. If we want to make it platform specific I think we'd need to amend the APIs to accept a "platform to entrypoint" mapping and store that with the registry metadata.

File: cli/module_registry.go
Comment (abe-winter) [reply cli/module_registry.go:658]: done in 595dafb2b

File: cli/module.schema.json
Comment (michaellee1019) [reply cli/module.schema.json]: Now that I am thinking more about the changes we need and the scope that was written, I think what I proposed in the scope isn't enough, but feel like it would be better to have a new section than overwrites.

Either you have the existing `entrypoint` and `build` sections or:

"environments": {
    "local": {"entrypoint":"path.sh"},
    "build:" {"linux/arm64":{"build":"build.sh","setup":"setup.sh","entrypoint":"run.sh"}},
}

A more significant change to the schema but puts us more in the right direction long term. If we use maps of maps and lists of maps everywhere we can easily add additional properties later. The fields we added early on don't have

Comment (abe-winter) [reply cli/module.schema.json]: re 'local' -- this applies in remote reloading as well. I'm open to a different name change though

re overwriting module_id -- remember this only applies in reloading mode. a user could just as easily edit meta.json by hand

yup resubmit; I changed where it lives (so that it can modify entrypoint) and the dev key

**Comment by abe-winter** [comment]
:grimacing: this needs more design I think -- closing for now

**Pull request closed without merging**
